nixos/ups: run upsmon as non-root user "nutmon" by default

yodaldevoid opened the pull request:
## Description of changes

Fixes #318582

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bjornfor commented:
@yodaldevoid: Do you have a working config with non-root user?

yodaldevoid commented:
> @yodaldevoid: Do you have a working config with non-root user?

Yes, you can use `lib.mkForce` to force the setting. e.g.
```
power.ups.upsmon.settings.RUN_AS_USER = lib.mkForce config.users.users.nutmonitor.name;
```

bjornfor commented:
> > @yodaldevoid: Do you have a working config with non-root user?
> 
> Yes, on stable you can use `lib.mkForce` to force the setting. e.g.
> 
> ```
> power.ups.upsmon.settings.RUN_AS_USER = lib.mkForce config.users.users.nutmonitor.name;
> ```

So there's no permission errors resulting from that? I assumed it was running as root for a reason (permissions). If it's that easy to run as non-root, why don't we create that "nutmonitor" user in NixOS by default and have everyone use that? What's the use case for users to be able to run NUT as arbitrary user?

yodaldevoid commented:
> So there's no permission errors resulting from that?
Nope, no issues. It even has access to a root-only password file at startup.

> I assumed it was running as root for a reason (permissions). If it's that easy to run as non-root, why don't we create that "nutmonitor" user in NixOS by default and have everyone use that? What's the use case for users to be able to run NUT as arbitrary user?
I can only make conjectures at this point, but I figure it was the easier thing at the time. If a user was generated and used, I know I would want a way to configure that (at least to turn it off to replace with my custom configured user). At that point, with how easy it is to generate users, why no just have the user do it themself.

bjornfor commented:
> If a user was generated and used, I know I would want a way to configure that (at least to turn it off to replace with my custom configured user). At that point, with how easy it is to generate users, why no just have the user do it themself.

That doesn't really explain a use case except for "because we can". What exactly do you want to customize? User name? Additional groups? (That can be done by extending/overriding a base user provided by NixOS.)

After reading https://networkupstools.org/docs/man/upsmon.conf.html (the RUN_AS_USER part), I think this patch is going in the wrong direction.

The way I think it should work is that NixOS generates a user (e.g. "nutmon" as suggested by upstream) and ensures the config file is readable for that user, so that reloading works. (This is actually how it works today, except that the generated user is "root".) Having to use lib.mkForce to change the user from under NixOS is a good thing; it implies a bit of "I know what I'm doing and accept the risks". Because this patch makes it easier to break the ownership of the config file.

bjornfor reviewed (dismissed):
I think we should fix the TODO in NixOS (for all people) instead of making it a bit easier to fix the issue out-of-tree only for expert users, see https://github.com/NixOS/nixpkgs/pull/318630#issuecomment-2292869018. (Expert users can just `lib.mkForce` the setting anyway.)

bjornfor commented:
I realized that we can turn this around; we can make RUN_AS_USER easily configurable, but then the permissions on the config files must automatically follow. I'm not a fan of making them easily diverge.

yodaldevoid committed:
nixos/ups: run upsmon as non-root user "nutmon" by default

NUT documentation strongly suggests running `upsmon` as a non-root user
with very limited permissions. This includes only read access to
`upsmon.conf`.

Previously the `RUN_AS_USER` setting could be overwritten with
`mkForce`, but this left the system with technically incorrect
permissions on `upsmon.conf` that would prevent reloading. This patch
provides a way to change the user that `upsmon` runs as while
maintaining correct permissions on `upsmon.conf`.

Fixes #318582

yodaldevoid commented:
I've finally got around to making this a more proper solution. I've followed the pattern I've seen other packages use; a default user is created if the user does not specify a different one. The same goes for this user's group.

bjornfor reviewed (commented):
Looks good. Didn't test. Commit message needs update ("run as non-root user 'nutmon'", or something?).

yodaldevoid commented:
I've updated the commit message with rationale for this change.

bjornfor commented:
Does shutdown work with this non-root?

bjornfor commented:
Thanks!

bjornfor commented on nixos/modules/services/monitoring/ups.nix:
```diff
@@ -382,7 +399,6 @@ let
         );
         NOTIFYCMD = lib.mkDefault "${pkgs.nut}/bin/upssched";
         POWERDOWNFLAG = lib.mkDefault "/run/killpower";
-        RUN_AS_USER = "root"; # TODO: replace 'root' by another username.
```
In the previous hunk, RUN_AS_USER is set to nutmon. Why isn't this set here too? Feels like doc and impl. disagrees.

yodaldevoid replied:
Ah, good catch. I missed that one when switching to the `-u` option.

bjornfor commented:
@yodaldevoid: I had plans to test this before merge, but it keep getting pushed back. So I think what we do now is that you say to me whether you think I should merge :-)

yodaldevoid commented:
@bjornfor I've been running this patch rebased on release-24.11 for a while on my personal machine with no issue, so I'd say merge it.

bjornfor commented:
@yodaldevoid: Thanks!
